sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/StatisticsSupport.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader;
+
+/**
+ * A mix in interface for `DataSourceV2Reader`. Users can implement this interface to propagate
+ * statistics like sizeInBytes, to Spark.
+ */
+public interface StatisticsSupport {
+  // todo: shall we add more statistics? what do we want?
+  long sizeInBytes();
```

Comment:
I think this should be more clear that the statistics returned should be for the reader as configured with partition pruning or predicate push-down. That would make this more useful for things like deciding whether a read should be used for a broadcast join.

I'd also include # of rows for the read, plus a few stats for each column: # values, # nulls, NDV, and some form of histogram (TBD).

Reply:
If we create a Statistics object and use it in the interface, then it becomes public and changing the Statistics object will break binary compatibility. I suggest we add some basic but useful statistics here via different methods, and create new interfaces if we wanna add more in the future.

Reply:
I don't think that's quite correct. In Java 8 we can add default methods which do not affect binary compatibility, if it's an interface. In all versions of Java, you can add a bean implementation which takes optional values.

Reply:
Or, put differently, I don't see what the difference is for binary compat between:

```
interface MyReader {
    OptionalLong getSize();
    OptionalLong getRows();
    OptionalLong getDistinctValues(String columnName);
}
```

```
interface MyReader {
    Statistics getStatistics();
}

interface Statistics {
    OptionalLong getSize();
    OptionalLong getRows();
    OptionalLong getDistinctValues(String columnName);
}
```

in terms of backcompat?

Reply:
but it would avoid interface bloat

Reply:
there's also the implementation detail that it's probably simpler for me to get you all the statistics than it is to compute them one at a time. E.g. if I store my statistics in a remote store, than if I want to implement the stats methods efficiently I need to write a cache for statistics so I only get them once (which changes if anyone adds another filter, so that's not so easy), whilst if I return a stats object I can return you an immutable object which is valid at the time it was created.

Reply:
+1